From: David Brownell <david-b@pacbell.net> 
Subject: fix hang in disable_periodic
References: bnc#403346
Mainline: unclear
Signed-off-by: Oliver Neukum <oneukum@suse.de>

The grand solution David proposed hasn't arrived. So we
use this to prevent usb_kill_urb() from causing a hang
in disable_periodic() under some circumstances.

--- linux-2.6.26/drivers/usb/host/ehci-sched.c.alt	2008-10-07 16:37:02.000000000 +0200
+++ linux-2.6.26/drivers/usb/host/ehci-sched.c	2008-10-07 16:47:03.000000000 +0200
@@ -467,6 +467,13 @@ static int disable_periodic (struct ehci
 	if (--ehci->periodic_sched)
 		return 0;
 
+	/* ehci->periodic_sched isn't an exact shadow of CMD_PSE;
+	* make sure it wasn't already cleared.
+	*/
+	cmd = ehci_readl(ehci, &ehci->regs->command);
+	if (!(cmd & CMD_PSE))
+		return 0;
+
 	/* did setting PSE not take effect yet?
 	 * takes effect only at frame boundaries...
 	 */
